File: plugins/llm/run-claude/repos/go-litellm/internal/router/router.go
// Package router is the live deployment registry (litellm.Router).
package router

import (
	"crypto/rand"
	"encoding/base64"
	"strings"
	"sync"
	"time"

	"github.com/noizu-labs/go-litellm/internal/config"
	"github.com/noizu-labs/go-litellm/internal/jsonx"
	"github.com/noizu-labs/go-litellm/internal/keys"
)

// Router holds the mutable model list and selects deployments.
type Router struct {
	mu          sync.RWMutex
	deployments []map[string]any
	cfg         *config.Config
	cooldowns   *CooldownCache
	Keys        *keys.Store
}

// BindSpec selects which deployments a key switch applies to.
type BindSpec struct {
	Target string // family ("zai") or exact model ("zai/opus")
	Prefix string // explicit prefix ("zai/")
	Using  string // current api_key_name
}

// New seeds from config.model_list.
func New(cfg *config.Config) *Router {
	ks := keys.New()
	ks.SeedFromEnv()
	r := &Router{cfg: cfg, cooldowns: NewCooldownCache(), Keys: ks}
	if cfg != nil {
		for _, d := range cfg.SnapshotModelList() {
			stored := ensureID(jsonx.Clone(d))
			if lp := jsonx.Nested(stored, "litellm_params"); lp != nil {
				ks.InferName(lp)
			}
			r.deployments = append(r.deployments, stored)
		}
	}
	return r
}

// Deployments returns a snapshot of the registry.
func (r *Router) Deployments() []map[string]any {
	r.mu.RLock()
	defer r.mu.RUnlock()
	out := make([]map[string]any, len(r.deployments))
	copy(out, r.deployments)
	return out
}

// Group returns deployments whose model_name matches (after alias resolution).
func (r *Router) Group(modelName string) []map[string]any {
	resolved := r.resolveAlias(modelName)
	var out []map[string]any
	for _, d := range r.Deployments() {
		if jsonx.Str(d, "model_name") == resolved {
			out = append(out, d)
		}
	}
	return out
}

// ModelNames returns distinct visible model_names.
func (r *Router) ModelNames() []string {
	seen := map[string]bool{}
	var out []string
	for _, d := range r.Deployments() {
		name := jsonx.Str(d, "model_name")
		if name == "" || seen[name] {
			continue
		}
		seen[name] = true
		out = append(out, name)
	}
	return out
}

// Select picks one deployment for a requested model.
func (r *Router) Select(modelName string) (map[string]any, bool) {
	candidates := r.Group(modelName)
	if len(candidates) == 0 {
		return nil, false
	}
	var available []map[string]any
	for _, d := range candidates {
		if !r.cooldowns.CooledDown(jsonx.Str(d, "model_id")) {
			available = append(available, d)
		}
	}
	pool := available
	if len(pool) == 0 {
		pool = candidates
	}
	return Pick(r.strategy(), pool), true
}

// Lookup is Select that returns nil when missing.
func (r *Router) Lookup(modelName string) map[string]any {
	d, ok := r.Select(modelName)
	if ok {
		return r.applyKey(d)
	}
	// Claude Code v2.1.116+ appends "[1m]"; groq-pro used to miss those aliases.
	if strings.HasSuffix(modelName, "[1m]") {
		if d, ok = r.Select(strings.TrimSuffix(modelName, "[1m]")); ok {
			return r.applyKey(d)
		}
	}
	return nil
}

// applyKey clones a deployment with the named key resolved into api_key.
func (r *Router) applyKey(d map[string]any) map[string]any {
	if d == nil || r.Keys == nil {
		return d
	}
	lp := jsonx.Nested(d, "litellm_params")
	if lp == nil {
		return d
	}
	k := r.Keys.Resolve(lp)
	if k == "" || k == jsonx.Str(lp, "api_key") {
		return d
	}
	out := jsonx.Clone(d)
	lp2 := jsonx.Clone(lp)
	lp2["api_key"] = k
	out["litellm_params"] = lp2
	return out
}

// BindKey sets api_key_name on matching deployments. Returns model_names updated.
func (r *Router) BindKey(spec BindSpec, keyName string) []string {
	keyName = keys.Canonical(keyName)
	r.mu.Lock()
	defer r.mu.Unlock()
	var updated []string
	for _, d := range r.deployments {
		name := jsonx.Str(d, "model_name")
		lp := jsonx.Nested(d, "litellm_params")
		if lp == nil {
			lp = map[string]any{}
			d["litellm_params"] = lp
		}
		match := false
		switch {
		case spec.Using != "":
			match = jsonx.Str(lp, "api_key_name") == keys.Canonical(spec.Using)
		case spec.Prefix != "":
			match = strings.HasPrefix(name, spec.Prefix)
		case spec.Target != "":
			match = keys.Matches(name, spec.Target)
		}
		if !match {
			continue
		}
		lp["api_key_name"] = keyName
		updated = append(updated, name)
	}
	return updated
}

// Bindings is the live model → key map (no secrets).
func (r *Router) Bindings() []map[string]any {
	var out []map[string]any
	for _, d := range r.Deployments() {
		lp := jsonx.Nested(d, "litellm_params")
		out = append(out, map[string]any{
			"model_name": jsonx.Str(d, "model_name"),
			"model_id":   jsonx.Str(d, "model_id"),
			"key":        jsonx.Str(lp, "api_key_name"),
		})
	}
	if out == nil {
		out = []map[string]any{}
	}
	return out
}

// Add registers a deployment and assigns a model_id if absent.
func (r *Router) Add(deployment map[string]any) map[string]any {
	stored := ensureID(jsonx.Clone(deployment))
	if _, ok := stored["litellm_params"]; !ok {
		stored["litellm_params"] = map[string]any{}
	}
	if _, ok := stored["model_info"]; !ok {
		stored["model_info"] = map[string]any{}
	}
	r.mu.Lock()
	r.deployments = append(r.deployments, stored)
	r.mu.Unlock()
	return stored
}

// Delete removes a deployment by model_id.
func (r *Router) Delete(modelID string) bool {
	r.mu.Lock()
	defer r.mu.Unlock()
	kept := r.deployments[:0]
	found := false
	for _, d := range r.deployments {
		if jsonx.Str(d, "model_id") == modelID {
			found = true
			continue
		}
		kept = append(kept, d)
	}
	if !found {
		return false
	}
	// re-slice into a new backing array so we don't leak the last elem
	out := make([]map[string]any, len(kept))
	copy(out, kept)
	r.deployments = out
	return true
}

// Update merges changes into a deployment by model_id.
func (r *Router) Update(modelID string, changes map[string]any) (map[string]any, bool) {
	r.mu.Lock()
	defer r.mu.Unlock()
	for i, d := range r.deployments {
		if jsonx.Str(d, "model_id") != modelID {
			continue
		}
		updated := jsonx.DeepMerge(d, changes)
		r.deployments[i] = updated
		return updated, true
	}
	return nil, false
}

// Set replaces the entire deployment set.
func (r *Router) Set(list []map[string]any) {
	seed := make([]map[string]any, 0, len(list))
	for _, d := range list {
		seed = append(seed, ensureID(jsonx.Clone(d)))
	}
	r.mu.Lock()
	r.deployments = seed
	r.mu.Unlock()
}

// CoolDown marks a deployment failed.
func (r *Router) CoolDown(modelID string, reason string) {
	r.cooldowns.Add(modelID, reason, r.cooldownTime())
}

// Cooldowns returns currently cooled-down model_ids.
func (r *Router) Cooldowns() []string {
	return r.cooldowns.Active()
}

// ClearCooldown removes a cooldown.
func (r *Router) ClearCooldown(modelID string) {
	r.cooldowns.Clear(modelID)
}

func (r *Router) resolveAlias(modelName string) string {
	if r.cfg == nil {
		return modelName
	}
	aliases := jsonx.AsMap(r.cfg.RouterSetting("model_group_alias"))
	if aliases == nil {
		return modelName
	}
	switch v := aliases[modelName].(type) {
	case string:
		return v
	default:
		if m := jsonx.AsMap(v); m != nil {
			if real := jsonx.Str(m, "model"); real != "" {
				return real
			}
		}
	}
	return modelName
}

func (r *Router) strategy() string {
	if r.cfg == nil {
		return "simple-shuffle"
	}
	if s, ok := r.cfg.RouterSetting("routing_strategy").(string); ok && s != "" {
		return s
	}
	return "simple-shuffle"
}

func (r *Router) cooldownTime() time.Duration {
	if r.cfg == nil {
		return 60 * time.Second
	}
	switch v := r.cfg.RouterSetting("cooldown_time").(type) {
	case int:
		return time.Duration(v) * time.Second
	case int64:
		return time.Duration(v) * time.Second
	case float64:
		return time.Duration(v * float64(time.Second))
	default:
		return 60 * time.Second
	}
}

func ensureID(d map[string]any) map[string]any {
	if info := jsonx.Nested(d, "model_info"); info != nil {
		if id := jsonx.Str(info, "id"); id != "" {
			d["model_id"] = id
			return d
		}
	}
	if id := jsonx.Str(d, "model_id"); id != "" {
		return d
	}
	buf := make([]byte, 16)
	_, _ = rand.Read(buf)
	id := base64.RawURLEncoding.EncodeToString(buf)
	d["model_id"] = id
	info := jsonx.Nested(d, "model_info")
	if info == nil {
		info = map[string]any{}
	}
	info["id"] = id
	d["model_info"] = info
	return d
}

// CooldownCache tracks per-deployment failure cooldowns.
type CooldownCache struct {
	mu   sync.Mutex
	data map[string]cooldown
}

type cooldown struct {
	expires time.Time
	reason  string
}

func NewCooldownCache() *CooldownCache {
	return &CooldownCache{data: map[string]cooldown{}}
}

func (c *CooldownCache) Add(modelID, reason string, d time.Duration) {
	if modelID == "" {
		return
	}
	c.mu.Lock()
	c.data[modelID] = cooldown{expires: time.Now().Add(d), reason: reason}
	c.mu.Unlock()
}

func (c *CooldownCache) CooledDown(modelID string) bool {
	if modelID == "" {
		return false
	}
	c.mu.Lock()
	defer c.mu.Unlock()
	cd, ok := c.data[modelID]
	if !ok {
		return false
	}
	if time.Now().Before(cd.expires) {
		return true
	}
	delete(c.data, modelID)
	return false
}

func (c *CooldownCache) Clear(modelID string) {
	c.mu.Lock()
	delete(c.data, modelID)
	c.mu.Unlock()
}

func (c *CooldownCache) Active() []string {
	c.mu.Lock()
	defer c.mu.Unlock()
	now := time.Now()
	var out []string
	for id, cd := range c.data {
		if now.Before(cd.expires) {
			out = append(out, id)
		}
	}
	return out
}

File: repos/ex-litellm/lib/ex_litellm/router/router.ex
defmodule ExLiteLLM.Router do
  @moduledoc """
  The deployment router — ex-litellm's `litellm.Router`.

  Holds the live model list as mutable state (seeded from the loaded
  `config.yaml`, then mutated at runtime by `/model/new|update|delete` — which is
  how run-claude registers models with `STORE_MODEL_IN_DB=True`). Responsibilities:

    * **registry** — the authoritative set of deployments, each with a stable
      `model_id`. `Deployments` reads through here.
    * **selection** — given a requested `model_name`, resolve `model_group_alias`,
      collect the matching group, filter out cooled-down deployments, and pick
      one per the configured routing strategy (default `simple-shuffle`).
    * **cooldowns** — on a deployment failure, cool it down for `cooldown_time`
      so selection skips it (via `ExLiteLLM.Router.CooldownCache`).

  A single GenServer owns the deployment list; selection reads a snapshot and
  applies the (pure) strategy so hot-path calls don't serialize on the server.
  """
  use GenServer

  alias ExLiteLLM.Config
  alias ExLiteLLM.Router.{CooldownCache, Strategy}

  @name __MODULE__

  # --- client API ---

  def start_link(opts \\ []) do
    GenServer.start_link(__MODULE__, opts, name: Keyword.get(opts, :name, @name))
  end

  @doc "All deployments currently registered."
  @spec deployments() :: [map()]
  def deployments, do: :persistent_term.get(pt_key(), [])

  @doc "Deployments in a model group (after alias resolution)."
  @spec group(String.t()) :: [map()]
  def group(model_name) do
    resolved = resolve_alias(model_name)
    Enum.filter(deployments(), &(&1["model_name"] == resolved))
  end

  @doc "Distinct visible model_names (hidden aliases excluded)."
  @spec model_names() :: [String.t()]
  def model_names do
    deployments()
    |> Enum.map(& &1["model_name"])
    |> Enum.reject(&is_nil/1)
    |> Enum.uniq()
  end

  @doc """
  Select one deployment for a requested model. Applies alias resolution,
  cooldown filtering, and the routing strategy. Returns `{:ok, deployment}` or
  `{:error, :no_deployment}`.
  """
  @spec select(String.t()) :: {:ok, map()} | {:error, :no_deployment}
  def select(model_name) do
    case group(model_name) do
      [] ->
        {:error, :no_deployment}

      candidates ->
        available = Enum.reject(candidates, &CooldownCache.cooled_down?(&1["model_id"]))
        pool = if available == [], do: candidates, else: available
        {:ok, Strategy.pick(strategy(), pool)}
    end
  end

  @doc "Register (add) a deployment. Assigns a model_id if absent. Returns the stored deployment."
  @spec add_deployment(map()) :: {:ok, map()}
  def add_deployment(deployment), do: GenServer.call(@name, {:add, deployment})

  @doc "Delete a deployment by model_id. Returns :ok or {:error, :not_found}."
  @spec delete_deployment(String.t()) :: :ok | {:error, :not_found}
  def delete_deployment(model_id), do: GenServer.call(@name, {:delete, model_id})

  @doc "Update a deployment by model_id (merges litellm_params/model_info)."
  @spec update_deployment(String.t(), map()) :: {:ok, map()} | {:error, :not_found}
  def update_deployment(model_id, changes), do: GenServer.call(@name, {:update, model_id, changes})

  @doc "Replace the entire deployment set (used when config reloads)."
  @spec set_deployments([map()]) :: :ok
  def set_deployments(list), do: GenServer.call(@name, {:set, list})

  @doc "Mark a deployment failed → cool it down."
  @spec cool_down(String.t(), term()) :: :ok
  def cool_down(model_id, reason), do: CooldownCache.add(model_id, reason, cooldown_time())

  # --- server ---

  @impl true
  def init(_opts) do
    seed = Config.get().model_list |> Enum.map(&ensure_id/1)
    publish(seed)
    {:ok, %{deployments: seed}}
  end

  @impl true
  def handle_call({:add, deployment}, _from, state) do
    stored = ensure_id(deployment)
    deployments = state.deployments ++ [stored]
    publish(deployments)
    {:reply, {:ok, stored}, %{state | deployments: deployments}}
  end

  def handle_call({:delete, model_id}, _from, state) do
    case Enum.split_with(state.deployments, &(&1["model_id"] == model_id)) do
      {[], _} ->
        {:reply, {:error, :not_found}, state}

      {_removed, kept} ->
        publish(kept)
        {:reply, :ok, %{state | deployments: kept}}
    end
  end

  def handle_call({:update, model_id, changes}, _from, state) do
    case Enum.find(state.deployments, &(&1["model_id"] == model_id)) do
      nil ->
        {:reply, {:error, :not_found}, state}

      dep ->
        updated = deep_merge(dep, changes)
        deployments = Enum.map(state.deployments, &if(&1["model_id"] == model_id, do: updated, else: &1))
        publish(deployments)
        {:reply, {:ok, updated}, %{state | deployments: deployments}}
    end
  end

  def handle_call({:set, list}, _from, state) do
    seed = Enum.map(list, &ensure_id/1)
    publish(seed)
    {:reply, :ok, %{state | deployments: seed}}
  end

  # --- helpers ---

  defp publish(deployments), do: :persistent_term.put(pt_key(), deployments)
  defp pt_key, do: {__MODULE__, :deployments}

  defp ensure_id(%{"model_info" => %{"id" => id}} = d) when is_binary(id) do
    Map.put(d, "model_id", id)
  end

  defp ensure_id(%{"model_id" => id} = d) when is_binary(id), do: d

  defp ensure_id(d) do
    id = 16 |> :crypto.strong_rand_bytes() |> Base.url_encode64(padding: false)

    d
    |> Map.put("model_id", id)
    |> Map.update("model_info", %{"id" => id}, &Map.put(&1, "id", id))
  end

  defp resolve_alias(model_name) do
    case Config.get().router_settings["model_group_alias"] do
      %{} = aliases ->
        case Map.get(aliases, model_name) do
          alias_target when is_binary(alias_target) -> alias_target
          %{"model" => real} -> real
          _ -> model_name
        end

      _ ->
        model_name
    end
  end

  defp strategy do
    Config.get().router_settings["routing_strategy"] || "simple-shuffle"
  end

  defp cooldown_time do
    case Config.get().router_settings["cooldown_time"] do
      n when is_number(n) -> round(n * 1000)
      _ -> 60_000
    end
  end

  defp deep_merge(base, changes) do
    Map.merge(base, changes, fn
      _k, %{} = v1, %{} = v2 -> deep_merge(v1, v2)
      _k, _v1, v2 -> v2
    end)
  end
end
